lua/telescope/pickers.lua

```diff
@@ -31,6 +32,188 @@ local ns_telescope_matching = a.nvim_create_namespace "telescope_matching"
 local ns_telescope_prompt = a.nvim_create_namespace "telescope_prompt"
 local ns_telescope_prompt_prefix = a.nvim_create_namespace "telescope_prompt_prefix"
 
+---@class telescope_popup_options
+---@field border table<1|2|3|4, integer>
+---@field borderchars table<1|2|3|4|5|6|7|8, string>
+---@field borderhighlight string
+---@field col integer
+---@field enter boolean
+---@field height integer
+---@field highlight string
+---@field line integer
+---@field minheight integer
+---@field title integer
+---@field titlehighlight integer
+---@field width integer
+
+-- Create three windows:
+-- 1. Prompt window
+-- 2. Options window
+-- 3. Preview window
+--
+---@param picker Picker
+local function default_create_layout(picker)
```

Comment:
i am thinking that we should maybe now move this in a new separate file

`lua/telescope/pickers/window.lua`

Reply:
> there are breaking changes regarding internals because status, ... changes,

Hmm 🤔 I tried to keep the status object same... only added an extra property `layout`.

<img width="1249" alt="image" src="https://github.com/nvim-telescope/telescope.nvim/assets/8050659/b664511d-4691-4f00-8552-6b81b03c3aa3">

If I broke compatibility, it wasn't intentional. Let me know if I missed something, If breaking changes can be avoided, I'd like to avoid it.

> i am going to run and play around with it for a couple of days and then do another quick review, see if i no longer like something, or whatever and then i am confident with merging that i think

Sounds good! 😃 👍🏼

> unless you still have open tasks on your list for this PR

I don't have anything else to implement. Just need to make the documentation better.

Reply:
> start picker with previewer disabled and then toggle it: ...er/start/telescope.nvim/lua/telescope/actions/layout.lua:32: attempt to index field 'preview' (a nil value)

Fixed it.

> start picker with previewer enabled, error: .../packer/start/plenary.nvim/lua/plenary/window/border.lua:177: attempt to index field '_border_win_options' (a nil value)

I can't repro this one. 🤔

Reply:
> If I broke compatibility, it wasn't intentional. Let me know if I missed something, If breaking changes can be avoided, I'd like to avoid it.

oh i missed that compat part. thats really good :+1: